### PR TITLE
drivers: spi: supporting Andes atcspi200 driver

### PR DESCRIPTION
**Pull request by kevinwang821020**
Support andes atcspi200 driver, and use board adp_xc7k_ae350 for testing.
Also the driver's Kconfig has fit the current policy, the Kconfig is depend on status in device node (DT_HAS_XXXXX).

### CODE REVIEW

File: drivers/spi/spi_andes_atcspi200.c [hunk 1]
```diff
@@ -0,0 +1,629 @@
...
+			  REG_IEN_END_MSK;
+	}
+
+	sys_write32(tctrl, SPI_TCTRL(dev));
+
+	/* Enable TX/RX FIFO interrupts */
+	sys_write32(int_msk, SPI_INTEN(dev));
+
+	if (!data->is_cmdaddr_mode) {
+		/* Start transfering */
```
Comment (fkokosinski) [inline drivers/spi/spi_andes_atcspi200.c:323]: `transfering` -> `transferring` (checkpatch complains about this typo)

File: dts/bindings/spi/andestech.atcspi200.yaml [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+include: spi-controller.yaml
+
+properties:
+
+    reg:
+      required: true
+
+    num-cs:
+      type: int
+      required: false
```
Comment (fkokosinski) [inline dts/bindings/spi/andestech.atcspi200.yaml:21]: IIRC `required` defaults to `false`, so there's no need to specify it if you don't want to make this field required.

File: drivers/spi/spi_andes_atcspi200.c (hunk 1 above)
Comment (kevinwang821020) [reply drivers/spi/spi_andes_atcspi200.c]: Thanks for reviewing!!
I have fixed it.

File: dts/bindings/spi/andestech.atcspi200.yaml (hunk 2 above)
Comment (kevinwang821020) [reply dts/bindings/spi/andestech.atcspi200.yaml]: Oh! I got it.
Thank you for your help, I have fixed it.

File: drivers/spi/spi_andes_atcspi200.c [hunk 3]
```diff
@@ -0,0 +1,629 @@
...
+	static struct spi_atcspi200_data spi_atcspi200_dev_data_##n = {	\
+		SPI_CONTEXT_INIT_LOCK(spi_atcspi200_dev_data_##n, ctx),	\
+		SPI_CONTEXT_INIT_SYNC(spi_atcspi200_dev_data_##n, ctx),	\
+		SPI_IF_NO_CMD(n)					\
+		SPI_BUSY_INIT						\
+		SPI_DMA_CHANNEL(n, rx, RX, PERIPHERAL, MEMORY)		\
+		SPI_DMA_CHANNEL(n, tx, TX, MEMORY, PERIPHERAL)		\
+	};								\
+	static void spi_atcspi200_cfg_##n(void);			\
+	static struct spi_atcspi200_cfg spi_atcspi200_dev_cfg_##n = {	\
```
Comment (bbilas) [inline drivers/spi/spi_andes_atcspi200.c:601]:
```suggestion
	static const struct spi_atcspi200_cfg spi_atcspi200_dev_cfg_##n = {	\
```

Comment (kevinwang821020) [reply drivers/spi/spi_andes_atcspi200.c:601]: Thanks for reviewing!

File: dts/bindings/spi/andestech.atcspi200.yaml [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+compatible: "andestech,atcspi200"
+
+include: spi-controller.yaml
+
+properties:
+
+    reg:
+      required: true
+
+    num-cs:
```
Comment (bbilas) [inline dts/bindings/spi/andestech.atcspi200.yaml:19]: Where is that being used?

**Comment by bbilas** [comment]
What about non-dedicated CS gpios? Is there any particular reason not to use them?

File: dts/bindings/spi/andestech.atcspi200.yaml (hunk 4 above)
Comment (kevinwang821020) [reply dts/bindings/spi/andestech.atcspi200.yaml]: Thanks for reviewing.

Originally, we just hoped to let users know the number of CS our spi driver support. And due to our spi IP provide dedicated CS(automatic control by hardware), so we only support one dedicated CS pin.

After seeing your suggestion, we also think that it is not very suitable to be a property just for the purpose of providing information, so we decided to remove it, thank you very much.

**Comment by kevinwang821020** [comment]
> What about non-dedicated CS gpios? Is there any particular reason not to use them?

**Comment by kevinwang821020** [comment]
> What about non-dedicated CS gpios? Is there any particular reason not to use them?

Thanks for your review.
Because the dedicated CS pin in our SPI IP is automatically controlled by hardware, it is more efficient.
So currently we tend to only support dedicated CS for our SPI driver.

**Comment by bbilas** [comment]
> > What about non-dedicated CS gpios? Is there any particular reason not to use them?
> 
> Thanks for your review. Because the dedicated CS pin in our SPI IP is automatically controlled by hardware, it is more efficient. So currently we tend to only support dedicated CS for our SPI driver.

That might be a problem in case of troubles with routing dedicated CS GPIO on the PCB design. In that case, the user will not be able to use non-dedicated CS GPIO instead as it is done in the rest of the SPI drivers in the Zephyr.

File: drivers/spi/spi_andes_atcspi200.c [hunk 5]
```diff
@@ -0,0 +1,629 @@
+/*
+ * Copyright (c) 2019 Andes Technology Corporation.
```
Comment (bbilas) [inline drivers/spi/spi_andes_atcspi200.c:2]: s/2019/2022

File: drivers/spi/Kconfig.andes_atcspi200 [hunk 6]
```diff
@@ -0,0 +1,12 @@
+# Kconfig Andes ATCSPI200 SPI configuration options
+#
+# Copyright (c) 2019 Andes Technology Corporation.
```
Comment (bbilas) [inline drivers/spi/Kconfig.andes_atcspi200:3]: s/2019/2022

File: dts/bindings/spi/andestech.atcspi200.yaml [hunk 7]
```diff
@@ -0,0 +1,16 @@
+#
+# Copyright (c) 2019, Andes Technology Corporation.
```
Comment (bbilas) [inline dts/bindings/spi/andestech.atcspi200.yaml:2]: s/2019/2022

File: boards/riscv/adp_xc7k_ae350/adp_xc7k_ae350_defconfig [hunk 8]
```diff
@@ -13,6 +13,9 @@ CONFIG_SERIAL=y
 # GPIO driver options
 CONFIG_GPIO=y
 
+# SPI driver options
+CONFIG_SPI=y
+
```
Comment (fkokosinski) [inline boards/riscv/adp_xc7k_ae350/adp_xc7k_ae350_defconfig:18]: Don't y-select `CONFIG_SPI` by default. See this PR: https://github.com/zephyrproject-rtos/zephyr/pull/48574

File: drivers/spi/spi_andes_atcspi200.c [hunk 9]
```diff
@@ -0,0 +1,629 @@
...
+#define REG_DATA        0x2c
+#define REG_CTRL        0x30
+#define REG_STAT        0x34
+#define REG_INTEN       0x38
+#define REG_INTST       0x3c
+#define REG_TIMIN       0x40
+#define REG_MCTRL       0x50
+#define REG_SLVST       0x60
+#define REG_SDCNT       0x64
+#define REG_CONFIG      0x7c
```
[the remark this answers is not in the document]
Comment (carlocaione) [reply drivers/spi/spi_andes_atcspi200.c:31]: uhm, too many of them for the DT. These are just offsets from a base address defined into the DT so I think this is fine.

File: drivers/spi/spi_andes_atcspi200.c [hunk 10]
```diff
@@ -0,0 +1,629 @@
...
+
+#define DEV_CFG(dev) ((const struct spi_atcspi200_cfg * const) (dev)->config)
+#define DEV_DATA(dev) ((struct spi_atcspi200_data *) (dev)->data)
```
Comment (carlocaione) [inline drivers/spi/spi_andes_atcspi200.c:34]: We are removing `DEF_CFG` and `DEV_DATA` from drivers. Do not use it.

File: drivers/spi/spi_andes_atcspi200.c [hunk 11]
```diff
@@ -0,0 +1,629 @@
...
+#define SPI_DATA(dev) (DEV_CFG(dev)->base + REG_DATA)
+#define SPI_CTRL(dev) (DEV_CFG(dev)->base + REG_CTRL)
+#define SPI_STAT(dev) (DEV_CFG(dev)->base + REG_STAT)
+#define SPI_INTEN(dev) (DEV_CFG(dev)->base + REG_INTEN)
+#define SPI_INTST(dev) (DEV_CFG(dev)->base + REG_INTST)
+#define SPI_TIMIN(dev) (DEV_CFG(dev)->base + REG_TIMIN)
+#define SPI_MCTRL(dev) (DEV_CFG(dev)->base + REG_MCTRL)
+#define SPI_SLVST(dev) (DEV_CFG(dev)->base + REG_SLVST)
+#define SPI_SDCNT(dev) (DEV_CFG(dev)->base + REG_SDCNT)
+#define SPI_CONFIG(dev) (DEV_CFG(dev)->base + REG_CONFIG)
```
Comment (carlocaione) [inline drivers/spi/spi_andes_atcspi200.c:51]: Please remove the macros not actually used.

File: drivers/spi/spi_andes_atcspi200.c [hunk 12]
```diff
@@ -0,0 +1,629 @@
...
+
+/* Field mask of SPI transfer format register */
+#define REG_TFMAT_CPHA_OFFSET (0)
+#define REG_TFMAT_CPOL_OFFSET (1)
+#define REG_TFMAT_SLVMODE_OFFSET (2)
+#define REG_TFMAT_LSB_OFFSET (3)
+#define REG_TFMAT_MOSI_BIDIR_OFFSET (4)
+#define REG_TFMAT_DATA_MERGE_OFFSET (7)
+#define REG_TFMAT_DATA_LEN_OFFSET (8)
+#define REG_TFMAT_ADDR_LEN_OFFSET (16)
```
Comment (carlocaione) [inline drivers/spi/spi_andes_atcspi200.c:61]: Pick one indentation style and stick to it (cfr with `REG_*`)

File: drivers/spi/spi_andes_atcspi200.c [hunk 13]
```diff
@@ -0,0 +1,629 @@
...
+/* Field mask of SPI transfer control register */
+#define REG_TCTRL_RD_TCNT_OFFSET (0)
+#define REG_TCTRL_DUMMY_CNT_OFFSET (9)
+#define REG_TCTRL_TOKEN_VAL_OFFSET (11)
+#define REG_TCTRL_WR_TCNT_OFFSET (12)
+#define REG_TCTRL_TOKEN_EN_OFFSET (21)
+#define REG_TCTRL_TRNS_MODE_OFFSET (24)
+#define REG_TCTRL_ADDR_FMT_OFFSET (28)
+#define REG_TCTRL_ADDR_EN_OFFSET (29)
+#define REG_TCTRL_CMD_EN_OFFSET (30)
```
Comment (carlocaione) [inline drivers/spi/spi_andes_atcspi200.c:88]: Remove unneeded defines and move everything to an header file.

File: drivers/spi/spi_andes_atcspi200.c [hunk 14]
```diff
@@ -0,0 +1,629 @@
...
+#define REG_STAT_TX_EMPTY_OFFSET (22)
+#define REG_STAT_TX_FULL_OFFSET (23)
+
+#define REG_STAT_SPI_ACT_MSK (0x1 << REG_STAT_SPI_ACT_OFFSET)
+#define REG_STAT_RX_NUM_MSK (0x1f << REG_STAT_RX_NUM_OFFSET)
+#define REG_STAT_RX_EMPTY_MSK (0x1 << REG_STAT_RX_EMPTY_OFFSET)
+#define REG_STAT_RX_FULL_MSK (0x1 << REG_STAT_RX_FULL_OFFSET)
+#define REG_STAT_TX_NUM_MSK (0x1f << REG_STAT_TX_NUM_OFFSET)
+#define REG_STAT_TX_EMPTY_MSK (0x1 << REG_STAT_TX_EMPTY_OFFSET)
+#define REG_STAT_TX_FULL_MSK (0x1 << REG_STAT_TX_FULL_OFFSET)
```
Comment (carlocaione) [inline drivers/spi/spi_andes_atcspi200.c:187]: Use `GENMASK` when possible.

File: drivers/spi/spi_andes_atcspi200.c [hunk 15]
```diff
@@ -0,0 +1,629 @@
...
+#define RX_FIFO_SIZE_SETTING(dev) \
+	(sys_read32(SPI_CONFIG(dev)) & REG_CONFIG_RX_FIFO_SIZE_MSK)
+#define RX_FIFO_SIZE(dev) \
+	(2 << (RX_FIFO_SIZE_SETTING(dev) >> REG_CONFIG_RX_FIFO_SIZE_OFFSET))
+
+#define TX_NUM_STAT(dev) (sys_read32(SPI_STAT(dev)) & REG_STAT_TX_NUM_MSK)
+#define RX_NUM_STAT(dev) (sys_read32(SPI_STAT(dev)) & REG_STAT_RX_NUM_MSK)
+#define GET_TX_NUM(dev) (TX_NUM_STAT(dev) >> REG_STAT_TX_NUM_OFFSET)
+#define GET_RX_NUM(dev) (RX_NUM_STAT(dev) >> REG_STAT_RX_NUM_OFFSET)
+
```
Comment (carlocaione) [inline drivers/spi/spi_andes_atcspi200.c:214]: empty space

File: drivers/spi/spi_andes_atcspi200.c [hunk 16]
```diff
@@ -0,0 +1,629 @@
...
+	/* Set data length */
+	data_len = SPI_WORD_SIZE_GET(config->operation) - 1;
+	CLR_MASK(SPI_TFMAT(dev), REG_TFMAT_DATA_LEN_MSK);
+	SET_MASK(SPI_TFMAT(dev), (data_len << REG_TFMAT_DATA_LEN_OFFSET));
+
+	/* Set SPI frame format */
+	if (config->operation & SPI_MODE_CPHA)
+		SET_MASK(SPI_TFMAT(dev), REG_TFMAT_CPHA_MSK);
+	else
+		CLR_MASK(SPI_TFMAT(dev), REG_TFMAT_CPHA_MSK);
```
Comment (carlocaione) [inline drivers/spi/spi_andes_atcspi200.c:263]: Missing `{`

File: drivers/spi/spi_andes_atcspi200.c [hunk 17]
```diff
@@ -0,0 +1,629 @@
...
+
+	SET_MASK(SPI_CTRL(dev), TX_FIFO_THRESHOLD << REG_CTRL_TX_THRES_OFFSET);
+	SET_MASK(SPI_CTRL(dev), RX_FIFO_THRESHOLD << REG_CTRL_RX_THRES_OFFSET);
+
+	return 0;
+}
+
+static int spi_transfer(const struct device *dev, uint32_t len)
+{
+
```
Comment (carlocaione) [inline drivers/spi/spi_andes_atcspi200.c:288]: empty space

**Review by carlocaione — CHANGES_REQUESTED** [review]

File: drivers/spi/spi_andes_atcspi200.c (hunk 5 above)
Comment (kevinwang821020) [reply drivers/spi/spi_andes_atcspi200.c]: Thanks for reviewing.

I have fixed this incorrect date.

File: drivers/spi/Kconfig.andes_atcspi200 (hunk 6 above)
Comment (kevinwang821020) [reply drivers/spi/Kconfig.andes_atcspi200:3]: Thanks for reviewing.

I have fixed this incorrect date.

File: dts/bindings/spi/andestech.atcspi200.yaml (hunk 7 above)
Comment (kevinwang821020) [reply dts/bindings/spi/andestech.atcspi200.yaml]: Thanks for reviewing.

I have fixed this incorrect date.

File: boards/riscv/adp_xc7k_ae350/adp_xc7k_ae350_defconfig (hunk 8 above)
Comment (kevinwang821020) [reply boards/riscv/adp_xc7k_ae350/adp_xc7k_ae350_defconfig]: OK! I got it!

Thank a lot for your remind

File: drivers/spi/spi_andes_atcspi200.c (hunk 10 above)
Comment (kevinwang821020) [reply drivers/spi/spi_andes_atcspi200.c]: Thanks for reviewing.

I have remove the inappropriate usage.

File: drivers/spi/spi_andes_atcspi200.c (hunk 11 above)
Comment (kevinwang821020) [reply drivers/spi/spi_andes_atcspi200.c]: OK! I got it!

Thanks a lot!

File: drivers/spi/spi_andes_atcspi200.c (hunk 12 above)
Comment (kevinwang821020) [reply drivers/spi/spi_andes_atcspi200.c]: I have change the indentation style.
Thanks for your remind.

File: drivers/spi/spi_andes_atcspi200.c (hunk 13 above)
Comment (kevinwang821020) [reply drivers/spi/spi_andes_atcspi200.c]: OK, I see!
Thanks for your remind.

File: drivers/spi/spi_andes_atcspi200.c (hunk 17 above)
Comment (kevinwang821020) [reply drivers/spi/spi_andes_atcspi200.c]: OK, I see!
Thanks for your remind.

File: drivers/spi/spi_andes_atcspi200.c (hunk 16 above)
Comment (kevinwang821020) [reply drivers/spi/spi_andes_atcspi200.c]: OK, I see!
Thanks for your remind.

File: drivers/spi/spi_andes_atcspi200.c (hunk 15 above)
Comment (kevinwang821020) [reply drivers/spi/spi_andes_atcspi200.c]: OK, I see!
Thanks for your remind.

File: drivers/spi/spi_andes_atcspi200.c (hunk 14 above)
Comment (kevinwang821020) [reply drivers/spi/spi_andes_atcspi200.c]: That's great!
Thanks for your suggestion.

**Comment by kevinwang821020** [comment]
In addition to modify according to your suggestions, due to the spi context API has been updated in v3.2, I also modified it according to the updated content.

Thanks a lot.

**Comment by kevinwang821020** [comment]
> > > What about non-dedicated CS gpios? Is there any particular reason not to use them?
> > 
> > 
> > Thanks for your review. Because the dedicated CS pin in our SPI IP is automatically controlled by hardware, it is more efficient. So currently we tend to only support dedicated CS for our SPI driver.
> 
> That might be a problem in case of troubles with routing dedicated CS GPIO on the PCB design. In that case, the user will not be able to use non-dedicated CS GPIO instead as it is done in the rest of the SPI drivers in the Zephyr.

Thanks for your suggestion.
It is a great point.
And we have let our driver support non-dedicated CS via gpio.

**Comment by kevinwang821020** [comment]
It seems like there are some problem when building and running qemu_x86.
But I thought there's nothing I can do in this PR to avoid them.
Is it reasonable to keep these errors here?

**Comment by bbilas** [comment]
> It seems like there are some problem when building and running qemu_x86. But I thought there's nothing I can do in this PR to avoid them. Is it reasonable to keep these errors here?

It seems to be related to the https://github.com/zephyrproject-rtos/zephyr/pull/49646.

**Comment by kevinwang821020** [comment]
> > It seems like there are some problem when building and running qemu_x86. But I thought there's nothing I can do in this PR to avoid them. Is it reasonable to keep these errors here?
> 
> It seems to be related to the #49646.

Yes, it seems like the same issue.
Thanks a lot!

**Review by fkokosinski — DISMISSED** [review]

File: drivers/spi/spi_andes_atcspi200.c (hunk 3 above)
Comment (bbilas) [reply drivers/spi/spi_andes_atcspi200.c:410]: It seems that you have forgotten about that.

Comment (kevinwang821020) [reply drivers/spi/spi_andes_atcspi200.c]: It my mistake, thanks for your remind.

**Review by carlocaione — APPROVED** [review]

**Review by fkokosinski — APPROVED** [review]

**Pull request merged**